hazelcast/src/main/java/com/hazelcast/instance/impl/ClusterTopologyIntentTracker.java

```diff
@@ -0,0 +1,138 @@
+     * <p>
+     *     A cluster with specified replica count 3 is starting up. This is expected to result in a series
+     *     of updates similar to the following:
+     *     <pre>{@code
+     *     (-1, 3, 0, 0)
+     *     (3, 3, 1, 0)
+     *     (3, 3, 1, 1)
+     *     (3, 3, 2, 1)
+     *     (3, 3, 2, 2)
+     *     (3, 3, 3, 2)
+     *     (3, 3, 3, 3)
+     *     }</pre>
+     * </p>
+     * <p>
+     *     Assuming user requests scaling up a running cluster of 3 members to 5, the following
+     *     updates are expected:
+     *     <pre>{@code
+     *     (3, 3, 3, 3)
+     *     (3, 5, 4, 3)
+     *     (5, 5, 4, 4)
+     *     (5, 5, 5, 4)
+     *     (5, 5, 5, 5)
+     *     }</pre>
+     * </p>
+     * Notice that actual updates may differ (eg duplicate notifications of intermediate states may be received).
+     *
+     * @param previousSpecifiedReplicaCount   previous specified replicas count
+     * @param currentSpecifiedReplicaCount    current specified replicas count
+     * @param readyReplicasCount              number of ready replicas
+     * @param currentReplicasCount            number of current replicas
+     *
+     * @see NodeExtension#isReady()
+     */
+    void update(int previousSpecifiedReplicaCount, int currentSpecifiedReplicaCount, int readyReplicasCount,
+                int currentReplicasCount);
+
+    ClusterTopologyIntent getClusterTopologyIntent();
+
+    /**
+     * Initialize explicitly the cluster topology intent.
+     * @param clusterTopologyIntent
+     */
+    void initializeClusterTopologyIntent(ClusterTopologyIntent clusterTopologyIntent);
+
+    /**
+     * Handle Hazelcast node shutdown with the given cluster topology intent.
+     * @param clusterTopologyIntent
+     */
+    void shutdownWithIntent(ClusterTopologyIntent clusterTopologyIntent);
+
+    /**
+     * @return {@code true} if this instance of {@code ClusterTopologyIntentTracker} is active and tracking
+     *         cluster topology changes in a managed context, otherwise {@code false}.
+     */
+    boolean isEnabled();
+
+    /**
+     * @return  the number of requested Hazelcast members in the cluster, as determined by the specification
+     *          that is managed by the runtime context. When running Hazelcast in a Kubernetes StatefulSet,
+     *          this corresponds to the value in {@code StatefulSetSpec.size}.
+     */
+    int getCurrentSpecifiedReplicaCount();
+
+    /**
+     * Notifies the {@link ClusterTopologyIntentTracker} that Hazelcast members list has changed.
+     */
+    void onMembershipChange();
+
+    /**
+     * @return  {@code true} when this tracker accepts events coalesced, exactly as received from Kubernetes API,
+     *          or {@code false} when coalesced events should be split into separate updates.
+     *
+     * @implNote see also {@code KubernetesClient$StsMonitor#onMessage} method for implementation details
+     */
+    boolean acceptsCoalescedEvents();
```

Comment:
I think we can add reasoning in commit message to Javadoc.

```
Reasoning: for switching to FROZEN state,
to is not necessary to unbundle events
received from kubernetes API during rolling
restart, so by avoiding that we skip switching
to ACTIVE cluster state in between member
restarts during "rollout restart"
```

Reply:
added in 3d3f4701